Autopilot/Src/gps.cpp

```diff
@@ -0,0 +1,456 @@
+// position fix you must also send one of the position fix rate commands below too.
+#define PMTK_SET_NMEA_UPDATE_1HZ "$PMTK220,1000*1F\r\n"
+#define PMTK_SET_NMEA_UPDATE_5HZ "$PMTK220,200*2C\r\n"
+#define PMTK_SET_NMEA_UPDATE_10HZ "$PMTK220,100*2F\r\n"
+
+// Position fix update rate commands.
+#define PMTK_API_SET_FIX_CTL_1HZ "$PMTK100,1000,0,0,0,0*1C\r\n"
+#define PMTK_API_SET_FIX_CTL_5HZ "$PMTK100,200,0,0,0,0*2F\r\n"
+// Can't fix position faster than 5 times a second!
+
+#define PMTK_SET_BAUD_57600 "$PMTK251,57600*2C\r\n"
+#define PMTK_SET_BAUD_9600 "$PMTK251,9600*17\r\n"
+
+//turn on VTG (velocity info) and GGA (positional info) only
+#define PMTK_SET_NMEA_OUTPUT_GGAVTG "$PMTK314,0,0,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0*28\r\n"
+// turn on ALL THE DATA
+#define PMTK_SET_NMEA_OUTPUT_ALLDATA "$PMTK314,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0*28\r\n"
+// turn off output
+#define PMTK_SET_NMEA_OUTPUT_OFF "$PMTK314,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0*28\r\n"
+
+// to generate your own sentences, check out the MTK command datasheet and use a checksum calculator
+// such as the awesome http://www.hhhh.org/wiml/proj/nmeaxor.html
+
+// one of the 2 DGPS modes
+#define PMTK_ENABLE_SBAS "$PMTK313,1*2E\r\n"
+#define PMTK_ENABLE_WAAS "$PMTK301,2*2E\r\n"
+
+// ask for the release and version
+#define PMTK_Q_RELEASE "$PMTK605*31\r\n"
+
+// request for updates on antenna status
+#define PGCMD_ANTENNA "$PGCMD,33,1*6C\r\n"
+#define PGCMD_NOANTENNA "$PGCMD,33,0*6D\r\n"
+
+// https://nmeachecksum.eqth.net --> get checksums
+#define PUBX_CONFIG_NMEA "$PUBX,41,1,07,03,9600,0*10\r\n" //Change hex numbers later
+#define PUBX_SET_GGA "$PUBX,40,GGA,0,1,0,0,0,0*5B\r\n"
+#define PUBX_SET_VTG "$PUBX,40,VTG,0,1,0,0,0,0*5F\r\n"
+#define PUBX_SET_RMC_OFF "$PUBX,40,RMC,0,0,0,0,0,0*47\r\n"
+#define PUBX_SET_GSA_OFF "$PUBX,40,GSA,0,0,0,0,0,0*4A\r\n"
+#define PUBX_SET_GNS_OFF "$PUBX,40,GNS,0,0,0,0,0,0*41\r\n"
+#define PUBX_SET_GLL_OFF "$PUBX,40,GLL,0,0,0,0,0,0*5C\r\n"
+
+#define GPS_GGA_MESSAGE "GNGGA"
+#define GPS_VTG_MESSAGE "GNVTG"
+
+extern UART_HandleTypeDef huart4;
+
+NEOM8* NEOM8::gpsInstance = nullptr;
+
+NEOM8* NEOM8::GetInstance() {
+	if (!gpsInstance) {
+		gpsInstance = new NEOM8 {};
+	}
+
+	return gpsInstance;
+}
+
+NEOM8::NEOM8() : gpsData {},
+				 configured {false},
+				 dataAvailable {false},
+				 uartDataIsNew {false} {
+
+//	const uint8_t CFG_NMEA[16] = { 0x17, 0x20, 0b00011000, 0x40, 0x08, 0x01, 0x00, 0x00, 0x00, 0b01110110, 0x00, 0x01, 0x01, 0x01, 0x00, 0x00 };
+//	HAL_UART_Transmit_DMA(&huart4, (uint8_t*) CFG_NMEA, sizeof(CFG_NMEA));
+//
+//	HAL_UART_Transmit_DMA(&huart4, (uint8_t*) PUBX_CONFIG_NMEA, sizeof(PUBX_CONFIG_NMEA));
+//
+//	//	const uint8_t CFG_UART[21] = { 0x06, 0x00, 0x01, 0x00, 0b00000000, 0b00000000, 0x00, 0x00, 0b11000000, 0b00001000, 0x00, 0x01, 0b11000010, 0x00, 0x00, 0b00000010, 0x00, 0b00000010, 0x00, 0x00, 0x00 };
+//	//	HAL_UART_Transmit_DMA(&huart4, CFG_UART, sizeof(CFG_UART));
```

Comment:
Is all this stuff meant to be attempts at configuring the gps ? It's fine to keep it for the next guy, but walls of comments like this are terribly confusing. What you should do is surround the section with something like

#if 0
#endif

and just leave a line or two of comments above that explaining what this could be useful for.

Reply:
kk